Docs 5646 improve navigability of policies section pa (#140)
mulesoft/docs-api-manager: modules/ROOT/pages/policies-policy-level.adoc

= Policy Level
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: policy, custom, ootb, offline

> Policies implement an additional level of granularity with SLA tiering. Policies control access based on the following access structuring:

Reviewer: This seems confusing, we are saying that it adds a level of granularity with SLA tiering and then explain the different options which are not reduced to SLA based. Shouldn't we remove the "with SLA tiering"? and say that policies can implement an additional level of granularity and then list the options?

Author: Good catch! Done! Thanks!

* Resource-level policies

> Impose a limit on all requests that match the resource to which it is applied. All policies, except the Cross-Origin Resource Sharing (CORS) policy, are resource-level.

Reviewer: It doesn't impose a limit (behavior depends on the policy applied), it applies the policy on all requests that match.

Author: Done! Thanks!

> * SLA-based policies

Reviewer: Is SLA based a policy level granularity or a configuration level? because it doesn't change where the policy is applied, just the configuration ( a.k.a limit)

Reviewer: We can discuss this one offline.

Author: Ok! Thanks! Made a note!

Author: Done! Thanks!

Impose a limit specified by the level of access granted to the requesting application. The Rate Limiting policy is SLA-based.

When the API Gateway is configured to work as a Mule Cluster, the policy can be configured to work in a distributed manner.
  

>   [PA: This seems out of place and does not make sense.]

Reviewer: Yes, I don't think it goes here. We should just limit this to resource level policy that is the level of granularity that we add on top of API level (that is the default)

Author: Removed the sentence! Thanks!

> [PA: We need to tell the customers how to work with SLA-tiers for policies right?]

Reviewer: SLA tiers for me should go somewhere else and link to all the documentation on contracts, request api access and those flows.

Author: Made a note for our discussion! Thanks!

Author: Done! Thanks!

== Resource-Level Policies

Resource level policy support includes RAML and, since Mule 4, HTTP-based APIs. A policy applied at a resource level affects all HTTP methods (eg PUT, GET, POST), or selected HTTP methods, within the resource.

image::methods-resources.png[]

You can apply multiple conditions to filter your resources and HTTP methods. The URI template regex can be applied to one, many or all the methods that your API has.

IMPORTANT: You cannot apply the CORS policy to resources.

=== Usage Scenarios

The uses for resource level policies are limited mainly by your imagination. Here are a few possibilities:

* Applying policies to specific resources
* Securing a subset of an API
* Setting different limits on resources

If you have a RAML/OAS spec attached to your API, click Preview Resource Matching to check which resources your filters affect.

=== Concepts

Example API:

image::users-api-cropped.png[API Console displays resources and methods,height=329,width=732]

> A resource level policy supports Java regular expressions. For example, you can use the wildcard to apply a policy to multiple resources. When you apply the policy to the API, specify the resources to which it applies.

Reviewer: @balbifm were there differences between mule 3 and mule 4 in the expressions? if there are we should highlight them here

Reviewer 2: @priyaayyar contact me for more info on this

Author: Will do! Thanks!

----
 /api/users/.*/.* →
   /api/users/{userId}/assignments;
   /api/users/{userId}/assignments/{param}/values;
   /api/users/{userId}/permissions;
   /api/users/{userId}/permissions/{permissionId}
----

First, you apply a rate limiting policy to specific methods and resources, selecting the GET (read) method and using the following regular expression:

----
/api/users/.*/.*
----

You then configure the policy to limit requests to 100 requests per 1 hour, for example.

Next, you apply the rate limiting policy to specific methods and resources again, but selecting the POST (create) method and using the same regular expression as before. Configure 50 requests per 1 hour.

Finally, you can apply the rate limiting policy again to specific methods and resources, selecting the DELETE method. This time, you use and the same regular expression as before. For example, you can configure 25 requests per 2 hours.

> == Policies and SLA Tiers

Reviewer: This for me should go in a different section. Maybe we can review with you and Chirag to confirm, but it puts two important aspects in a long document and it might get lost.

Author: Made a note! Thanks!

You create the following tiers of access that limit apps calling the API to 3 or 5 requests, depending on the SLA level of the app.

* A tier that limits the requests to three per minute. No approval required.
* A tier that limits requests to five per minute and requires API Versions Owner approval of the app that wants to access the API.

In this procedure, you limit access to one API resource. Access to other resources are unlimited. When an app attempts to consume the protected resource, the policy is enforced. The request needs to include the expected user name/password. Repeated calls within SLA limits from the app to the API succeed; others fail.

> == SLA-Level Policies

Reviewer: Isn't this a duplicate from the above? it has the exact same text.

Author: Removed! Thanks!

= Policy Level
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: policy, custom, ootb, offline, resource level

> By default, policies are applied to the entire API. However, you can implement an additional level of granularity for policies, where access is controlled based on the a criteria, called resource-level policies. 

Reviewer 3: Clarify WHAT criteria. "...supplied criteria"?
```suggestion
By default, policies are applied to the entire API. However, you can implement an additional level of granularity for policies, one in which access is controlled based on the criteria. These are called _resource-level policies_. 
```

Author: Done! Thanks!

> In this level of granularity, policies are applied to only those request that match the criteria.  All policies, except the Cross-Origin Resource Sharing (CORS) policy, are resource-level.

Reviewer 3: ```suggestion
At the resource level of granularity, policies are applied to only those requests that match the criteria.  All policies, except the Cross-Origin Resource Sharing (CORS) policy, are resource-level.
```
Clarify: If all but CORS are resource level, then why state that all are by default applied to the whole API? This sounds contradictory.

Reviewer: All policies unless configured otherwise are applied to the entire API. The customer might decide to have more granularity by configuring it defining the resource level. This configuration is not available on CORS.

Author: Done! Thanks!

//* SLA-based policies
//Impose a limit specified by the level of access granted to the requesting application.The Rate Limiting policy is SLA-based.
// Is SLA based a policy level granularity or a configuration level? because it doesn't change where the policy is applied, just the configuration ( a.k.a limit) - Eva??

> 

Reviewer 3: FYI, I should not be editing a file that has questions outstanding to your SMEs. I am supposed to be editing content that has been technically reviewed and resolved.

Author: noted.

//[PA: We need to tell the customers how to work with SLA-tiers for policies right?]
//SLA tiers for me should go somewhere else and link to all the documentation on contracts, request api access and those flows. - Eva?

> Resource level policy support includes RAML and HTTP-based APIs (since Mule 4). A policy applied at a resource level affects all HTTP methods, for example, PUT, GET, POST, or selected HTTP methods, within the resource.

Reviewer 3: ```suggestion
Beginning with Mule 4, resource-level policy support includes RAML and HTTP-based APIs. A policy applied at a resource level affects all HTTP methods within the resource: PUT, GET, POST, and selected HTTP methods:
```
colon points to the image that follows, which should illustrate this sentence.

Reviewer: Is it clear in this sentence that RAML based APIs are supported in Mule 3?

Author: Is this clearer: "With Mule 4, resource-level policy support now includes HTTP-based APIs along with RAML APIs, as opposed to support for only RAML APIs in Mule 3." Done! Thanks!

image::methods-resources.png[]

> You can apply multiple conditions to filter your resources and HTTP methods. The URI template regex can be applied to one, many, or all the methods that your API has.

Reviewer 3: ```suggestion
You can apply multiple conditions to filter your resources and HTTP methods. The URI template regex can be applied to one, many, or all the methods in your API.
```

Author: Done! Thanks!

== Usage Scenarios

> The uses for resource level policies are limitless. Listed below are a few examples of how you might implement the resource-level policies:

Reviewer 3: ```suggestion
The uses for resource-level policies are limitless; here are a few:
```

Author: Added, "a few examples are".

* Applying policies to specific resources
* Securing a subset of an API
* Setting different limits on resources

> If you have a RAML/OAS spec attached to your API, click Preview Resource Matching to check which resources your filters affect.

Reviewer 3: GLOBALLY replace / (and/or) with the better of "and" or "or."
```suggestion
If you have a RAML or OAS spec attached to your API, click *Preview Resource Matching* to check which resources your filters affect.
```
WHERE, in what UI location, does the user click this control?

Author: Added a note as I have no idea where. Bane of working with preexisting content. Done! Thanks!

> Example API:

Reviewer 3: Can we identify this a little more specifically? "Following is an example API for the usage scenario"?

Author: Done! Thanks!

image::users-api-cropped.png[API Console displays resources and methods,height=329,width=732]

> A resource level policy supports Java regular expressions. For example, you can use the wildcard to apply a policy to multiple resources. When you apply the policy to the API, specify the resources to which it applies.

Reviewer 3: Is the second sentence really an example of the first?
```suggestion
A resource-level policy supports Java regular expressions. For example, you can use the wildcard to apply a policy to multiple resources. When you apply the policy to the API, specify the resources to which it applies:
```

Author: Done! Thanks!

// were there differences between mule 3 and mule 4 in the expressions? if there are we should highlight them here -- Fede Balbi?

image::rlp-conditions.png[]

> You can apply multiple conditions to filter your resources and methods. The URI template regex can be applied to one, many, or all the methods that your API has.

Reviewer 3: Do we need to repeat this paragraph?

Author: Good catch. Removed! Done! Thanks!

> Do _not_ use a placeholder, such as {userid}, in the regular expression. Using a placeholder in an expression fails because the placeholder does not match the actual node. In the case of the example placeholder {userid}, the node actually looks something like this:

Reviewer 3: WHY are we illustrating what we do not want our users to do? That seems very dangerous, based on how careless readers can be. I suggest limiting the following text to the first sentence only, and omitting the rest, as well as the image.
```suggestion
Do not use a placeholder, such as {userid}, in the regular expression. 
```

Author: Removed some and kept some as I thought worked. Done! Thanks!

`/api/users/671962fc-f076-4b19-bc38-45ba3a4e4095/permissions/1234`

* 671962fc-f076-4b19-bc38-45ba3a4e4095 is the user ID

> * 1234 is the ID of a permission

Reviewer 3: omit

Author: I think this is good to have. Done! Thanks!

> To apply a policy to resource `/api/users/{userid}` that represents a single user and all nodes below the resource, use the following expression:

Reviewer 3: ```suggestion
To apply a policy to resource `/api/users/{userid}` that represents a single user and all resource nodes, use the following expression:
```

Author: Done! Thanks!

= Policy Level
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> :keywords: policy, custom, ootb, offline

Reviewer: shouldn't we include resource level as keyword here?

Author: Good point! Thanks! Done!

> Policies implement an additional level of granularity, where access is controlled based on the a criteria, called resource-level policies. 

Reviewer: Maybe we should say before that by default policies are applied to the entire API unless more granularity is needed?

Author: Done! Thanks!

In this level of granularity, policies are applied to only those request that match the criteria.  All policies, except the Cross-Origin Resource Sharing (CORS) policy, are resource-level.

//* SLA-based policies
//Impose a limit specified by the level of access granted to the requesting application.The Rate Limiting policy is SLA-based.
// Is SLA based a policy level granularity or a configuration level? because it doesn't change where the policy is applied, just the configuration ( a.k.a limit) - Eva??

//[PA: We need to tell the customers how to work with SLA-tiers for policies right?]
//SLA tiers for me should go somewhere else and link to all the documentation on contracts, request api access and those flows. - Eva?

> Resource level policy support includes RAML and HTTP-based APIs (since Mule 4). A policy applied at a resource level affects all HTTP methods, for example, PUT, GET, POST, or selected HTTP methods, within the resource.

Reviewer: Is it clear that the clarification is only for HTTP based APIs?

Author: I think and if not my editor will tell me once this goes through edit.

image::methods-resources.png[]

You can apply multiple conditions to filter your resources and HTTP methods. The URI template regex can be applied to one, many, or all the methods that your API has.

== Usage Scenarios

The uses for resource level policies are limitless. Listed below are a few examples of how you might implement the resource-level policies:

* Applying policies to specific resources
* Securing a subset of an API
* Setting different limits on resources

If you have a RAML/OAS spec attached to your API, click Preview Resource Matching to check which resources your filters affect.

> // == Concepts

Reviewer: Are all examples excluded?

Author: Added back some!
